### PR TITLE
fix stuff

### PR DESCRIPTION
**Pull request by CandyCat1029**
fix the ca not placing when others have absorbtion and make it not placing crystals that hurt us more while cant kill other also fixed the antisuicide

### CODE REVIEW

**Comment by CandyCat1029** [comment]
when the damage - 0.5 is smaller than health damage could be larger than ur health which leads to suicide like when the dmg is 4 and ur health is 3.9 if 4-0.5 > 3.9 then itll continue but ull die from that also that one means that if selfdmg isnt larger than the enemy damage and also the enemy damage isnt enough to kill him itll continue.

**Comment by CandyCat1029** [comment]
i made it to self + 0.5 so that the ca will at least remains 0.5 health for the user

**Comment by CandyCat1029** [comment]
and the damage calculator minus the absorbtion amount so if the absorbtion amount is 14 and crystal damage is  8 it wont place

**Comment by cookiedragon234** [comment]
🧐

**Comment by CandyCat1029** [comment]
if u use kami when the damage isnt enough  to damage all absorption amount it just wont place

**Comment by CandyCat1029** [comment]
cuz the damage is something like 14 but u got like 16 absorption amount and it will make the damage to be 0 and not accepting that spot

**Comment by CandyCat1029** [comment]
one crystal deals at most 16 damage to player when they have resistance and wear full prot with leggings blast prot and i mean when its in this situation

**Comment by CandyCat1029** [comment]
image

**Comment by CandyCat1029** [comment]
so kami ca wont do anything when u have just eaten a gap

**Comment by cookiedragon234** [comment]
Then find another solution, but incorrectly calculating the damage is not the right solution

**Comment by CandyCat1029** [comment]
also the calculation doesnt matter there it just make it only calculate the damage to the red health bar

**Comment by CandyCat1029** [comment]
it selects either the damage to red health bar or 0

**Comment by DarkiBoi** [comment]
👃

**Comment by iBuyMountainDew** [comment]
This PR remains a **bruh moment**. Fix your 'fix'.

**Comment by DarkiBoi** [comment]
^

**Comment by CandyCat1029** [comment]
in that code if (entity instanceof EntityPlayer) {
            EntityPlayer ep = (EntityPlayer) entity;
            DamageSource ds = DamageSource.causeExplosionDamage(explosion);
            damage = CombatRules.getDamageAfterAbsorb(damage, (float) ep.getTotalArmorValue(), (float) ep.getEntityAttribute(SharedMonsterAttributes.ARMOR_TOUGHNESS).getAttributeValue());

            int k = EnchantmentHelper.getEnchantmentModifierDamage(ep.getArmorInventoryList(), ds);
            float f = MathHelper.clamp(k, 0.0F, 20.0F);
            damage = damage * (1.0F - f / 25.0F);

            if (entity.isPotionActive(Potion.getPotionById(11))) {
                damage = damage - (damage / 4);
            }

            damage = Math.max(damage "" - ep.getAbsorptionAmount() "", 0.0F);
            return damage;
        }
static float calculateDamage(double posX, double posY, double posZ, Entity entity) {
        float doubleExplosionSize = 6.0F * 2.0F;
        double distancedsize = entity.getDistance(posX, posY, posZ) / (double) doubleExplosionSize;
        Vec3d vec3d = new Vec3d(posX, posY, posZ);
        double blockDensity = entity.world.getBlockDensity(vec3d, entity.getEntityBoundingBox());
        double v = (1.0D - distancedsize) * blockDensity;
        float damage = (float) ((int) ((v * v + v) / 2.0D * 7.0D * (double) doubleExplosionSize + 1.0D));
        double finald = 1;
        if (entity instanceof EntityLivingBase) {
            // we pass null as the exploder here
            //noinspection ConstantConditions
            finald = getBlastReduction((EntityLivingBase) entity, getDamageMultiplied(damage), new Explosion(mc.world, null, posX, posY, posZ, 6F, false, true)); that damage returned in the method above 
        }
        return (float) finald; then it is returned
    }
its returning the damage of that crystal minus the absorption amount or zero for the damage and the damage will be returned as a float in calculateDamage method
therefore the damage calculation only calculate the damage to the red heart bar because it minus the yellow heart bar damage out

**Comment by S-B99** [comment]
this is a bruh moment

**Comment by CandyCat1029** [comment]
if u just dont believe go make a thingy show the damage after placing and try it on a player with full armor and just eaten a gap kami ca just dont do anything cause Math.max(15 - 16 = -1, 0.0F) will set the damage to 0 and calculateDamage method will return 0 yet it actually deals 15 damage but just not on the red health

**Comment by S-B99** [comment]
> therefore the damage calculation only calculate the damage to the red heart bar because it minus the yellow heart bar damage out

and this is cookies point. you're intentionally making the damage calculation inaccurate

**Comment by CandyCat1029** [comment]
no i mean
it minus out the damage to the yellow heart bar thats why i remove the ep.getAbsorptionAmount()

**Comment by CandyCat1029** [comment]
it actually shouldnt be Math.max(damage, 0.0F) but the sentence should be removed

**Comment by S-B99** [comment]
yes, and as a result you're making the damage calculations inaccurate.

**Comment by CandyCat1029** [comment]
private static float getBlastReduction(EntityLivingBase entity, float damage, Explosion explosion) {
        if (entity instanceof EntityPlayer) {
            EntityPlayer ep = (EntityPlayer) entity;
            DamageSource ds = DamageSource.causeExplosionDamage(explosion);
            damage = CombatRules.getDamageAfterAbsorb(damage, (float) ep.getTotalArmorValue(), (float) ep.getEntityAttribute(SharedMonsterAttributes.ARMOR_TOUGHNESS).getAttributeValue());

            int k = EnchantmentHelper.getEnchantmentModifierDamage(ep.getArmorInventoryList(), ds);
            float f = MathHelper.clamp(k, 0.0F, 20.0F);
            damage = damage * (1.0F - f / 25.0F);

            if (entity.isPotionActive(MobEffects.RESISTANCE)) {
                damage = damage - (damage / 4);
            }

            return damage;
        }
        damage = CombatRules.getDamageAfterAbsorb(damage, (float) entity.getTotalArmorValue(), (float) entity.getEntityAttribute(SharedMonsterAttributes.ARMOR_TOUGHNESS).getAttributeValue());
        return damage;
    }

**Comment by CandyCat1029** [comment]
like this
it isnt inaccurate

**Comment by CandyCat1029** [comment]
cause the damage isnt adding the absorption amount

**Comment by CandyCat1029** [comment]
so no point for minus it

**Comment by iBuyMountainDew** [comment]
(≖ლ≖๑ )ﾌ

**Comment by CandyCat1029** [comment]
also if its adding the absorptionamount wots the point of doing Math.max(damage - ep.getAbsorptionAmount(), 0.0f)?
becuz it could be negative after the damage minusing the absorption amount
and its not logical for crystal damage to be negative

**Comment by CandyCat1029** [comment]
therefore the ep.getAbsorptionAmount() is wrong and not adding the damage dealt to absorption

**Comment by CandyCat1029** [comment]
image
kami crystal is on but not doing anything because 15-16 is -1 so damage change to 0 but instead should be 15

File: src/main/java/me/zeroeightsix/kami/module/modules/combat/CrystalAura.java [hunk 1]
```diff
@@ -296,10 +296,9 @@ public static float getBlastReduction(EntityLivingBase entity, float damage, Exp
             damage = damage * (1.0F - f / 25.0F);
 
             if (entity.isPotionActive(Potion.getPotionById(11))) {
-                damage = damage - (damage / 4);
+                damage = damage - (damage / 5);
```
Comment (CandyCat1029) [inline src/main/java/me/zeroeightsix/kami/module/modules/combat/CrystalAura.java:299]: should be 20% not 25%

**Comment by iBuyMountainDew** [comment]
You there mate? @CandyCat1029

**Comment by zeroeightysix** [comment]
I haven't taken a good look at this considering the conversation, is this ready for review? Or is it still 'wrong'?

**Comment by cookiedragon234** [comment]
@zeroeightysix it’s not ready to review yet, it’s a flawed idea

**Comment by CandyCat1029** [comment]
maybe i should really make a debug to show calculations and look at them in latest log to prove the calculations is right :P

**Comment by cookiedragon234** [comment]
Except the calculations aren’t right as you are purposefully discounting absorption from them

**Comment by CandyCat1029** [comment]
as i said before, Math.max(damage - ep.getAbsorptionAmount(), 0.0f) is becuz of the damage - absorption amount may be lower than 0, and damage of a crystal lower than 0 is not logical

**Comment by CandyCat1029** [comment]
so it shouldnt be deducted

**Comment by CandyCat1029** [comment]
image
image
proof for it

**Comment by CandyCat1029** [comment]
first because it does damage to ur absorption amount so all ur absorption amount got deducted my that crystal
and the world didnt refresh but my health refreshed so the second hit it does 15.4 damage to me, which is right after i got no absorption amount left

**Comment by CandyCat1029** [comment]
so i wanna say that it only calculate the damage to the red health bar but should be calculating the damage also to absorption amount therefore damage - ep.getAbsorptionAmount() is unnecessary

**Comment by CandyCat1029** [comment]
and should be returning damage instead of making the damage to be 0 because 15.5-16 is -0.5 which with math.max(-0.5, 0.0f) and 0.0f is larger so returning 0 as the damage

**Comment by CandyCat1029** [comment]
Command.sendChatMessage("Dealing " + calculateDamage(crystal, mc.player) + " damage to self"); just add this to hitting section and u can see damage

**Comment by dominikaaaa** [comment]
How is that "proof"? Explain it, because all I can see are 2 images, one with a crystal and one without. What is happening to both of them?

**Comment by CandyCat1029** [comment]
kami ca exploded the crystal
and the chat show the damage only to the red heart but not the yellow one

**Comment by CandyCat1029** [comment]
i cant record a vid cuz my recording software is ded

**Comment by dominikaaaa** [comment]
I still don't see what those 2 images are supposed to demonstrate.

**Comment by CandyCat1029** [comment]
i show where i am to explode the crystal
which im in the crystal

**Comment by CandyCat1029** [comment]
and it does 0 damage on first hit cuz
it does normally 16 damage and absorption amount is 16
so 16 - 16 is 0

**Comment by CandyCat1029** [comment]
and second hit is when i dont have yellow heart cuz the first crystal blew em up

**Comment by CandyCat1029** [comment]
and it shows 15.5
cuz the damage to red health is 15.5

**Comment by CandyCat1029** [comment]
i wanna say that deducting the ep.getAbsorptionAmount() makes it ignore damage to the yellow heart bar so its unnecessary

**Comment by dominikaaaa** [comment]
I feel this can be closed for the reasons already listed by existing participants of this pull

**Pull request closed without merging**